#!/usr/bin/env cwl-runner

class: CommandLineTool

id: "cgp-irap"

label: "Dockerised iRAP for RNAseq data analysis"

cwlVersion: v1.0

doc: |
    ![build_status](https://quay.io/repository/wtsicgp/dockstore-cgp-irap/status)
    A Docker container of iRAP. See the [dockstore-cgp-irap](https://github.com/cancerit/dockstore-cgp-irap) website for more information.

dct:creator:
  "@id": "yaobo.xu@sanger.ac.uk"
  foaf:name: Yaobo Xu
  foaf:mbox: "yx2@sanger.ac.uk"

requirements:
  - class: DockerRequirement
    dockerPull: "quay.io/wtsicgp/dockstore-cgp-irap:0.1.3"

hints:
  - class: ResourceRequirement
    coresMin: 1 # works but long, 8 recommended
    ramMin: 4000
    outdirMin: 80000

inputs:
  reference:
    type: File
    doc: "reference genome file in gzipped fasta format"
    inputBinding:
      prefix: -R
      position: 1
      separate: true
      shellQuote: true

  gtf:
    type: File
    doc: "a tar ball of GTF references generated by iRAP, or a single GTF file path"
    inputBinding:
      prefix: -g
      position: 2
      separate: true

  species:
    type: string
    doc: "species name"
    inputBinding:
      prefix: -s
      position: 3
      separate: true

  raw_read_file:
    type:
      type: array
      items: File
      inputBinding:
        prefix: -r
        separate: true
        shellQuote: true
    doc: "raw read files [bam|fastq]"
    inputBinding:
      position: 4

  exp_name:
    type: string?
    default: 'irap_analysis'
    doc: "analysis experiment name, iRAP use it to name the output folder as well"
    inputBinding:
      prefix: -o
      position: 5
      separate: true

  config:
    type: File?
    doc: "the configuration file for a analysis."
    inputBinding:
      prefix: -c
      position: 6
      separate: true
      shellQuote: true

  irap_conf_string:
    type:
      - "null"
      - type: array
        items: string
        inputBinding:
          prefix: -i
          separate: true
          shellQuote: true
    doc: "additional irap config strings"
    inputBinding:
      position: 7

  conta_genomes:
    type:
      - "null"
      - type: array
        items: File
        inputBinding:
          prefix: -m
          separate: true
          shellQuote: true
    doc: "contamination genomes in gzipped fasta format"
    inputBinding:
      position: 8

outputs:
  result_data:
    type: File
    outputBinding:
      glob: $(inputs.exp_name).tar.gz

  result_md5:
    type: File
    outputBinding:
      glob: $(inputs.exp_name).tar.gz.md5

  result_log:
    type: File
    outputBinding:
      glob: $(inputs.exp_name).log

baseCommand: ["irap_wrapper.sh"]
